readme.md
THE MASTER
==========

A base for plugins and themes

This plugin allocates THEWPMASTER Class which brings handy functionalety and presets
for plugin and theme development in wordpress.

See [Plugin Page](https://github.com/Xiphe/THEMASTER) for details


Installation
------------

1. Download the [latest version](https://github.com/Xiphe/THEMASTER/archive/master.zip).
2. Extract the archive and upload plugin to the `/wp-content/plugins/` directory of your Wordpress installation.
3. Activate the plugin through the 'Plugins' menu in WordPress
4. You can now use Plugins and Themes that require THEMASTER.


Support
-------

I've written this project for my own needs so i am not willing to give
full support. Anyway, i am very interested in any bugs, hints, requests
or whatever. Please use the [github issue system](https://github.com/Xiphe/THEMASTER/issues)
and i will try to answer.


Special Thanks
--------------

The update mechanics are inspired* by:
**[automatic-theme-plugin-update](https://github.com/jeremyclark13/automatic-theme-plugin-update)**  
  by [Kaspars Dambis](kaspars@konstruktors.com)  
  Modified by [Jeremy Clark](http://clark-technet.com)  
  [Donate](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=73N8G8UPGDG2Q)

The FileSelect module is inspired* by:
**[SLT File Select](http://sltaylor.co.uk/wordpress/plugins/slt-file-select/)**  
  by [Steve Taylor](http://sltaylor.co.uk)  
  released under GNU GENERAL PUBLIC LICENSE Version 2  


*Inspired means that i have seen this **brilliant project**, started to use them
and than started to write my own version with the intention to just have the
functionality i really use.  
I got the idea of how things are working by analyzing and using the code of
it and i can't say where i used snippets from this project but they've
influenced my code strongly.


3rd Party
---------

This plugin uses and includes some scripts from other people.
Here is a list:

### Composer
coffeescript/coffeescript, css-crush/css-crush, duvanmonsa/php-query, 
firephp/firephp-core leafo/lessphp, ikimea/browser and some of my own packages.

### PHP
* Diverse Snippets from all around the World - added the author whenever possible.
* **MIMEType-Array from [MIME Types Class](http://www.phpclasses.org/phpmimetypeclass)**  

###JS
* **[animaterotatescale](https://github.com/zachstronaut/jquery-animate-css-rotate-scale/blob/master/jquery-animate-css-rotate-scale.js)**  
  Devloped by zachstronaut

* **[coloumnmanager](http://p.sohei.org/jquery-plugins/columnmanager/)**  

* **combobox**  
  From jQueryUI-Examples - modifyed by me.

* **[colorbox](http://www.jacklmoore.com/colorbox)**  
  (c) 2011 Jack Moore - jacklmoore.com  
  License: http://www.opensource.org/licenses/mit-license.php

* **[jquery.cookie](https://github.com/carhartl/jquery-cookie)**  
  Copyright 2013 Klaus Hartl  
  [MIT License](https://github.com/carhartl/jquery-cookie/blob/master/MIT-LICENSE.txt)

* **[jquery-json](code.google.com/p/jquery-json)**  
  by Brantley Harris  
  [MIT License](http://opensource.org/licenses/mit-license.php)

* **[jQuery Mouse Wheel Plugin](https://github.com/brandonaaron/jquery-mousewheel)**  
  by Brandon Aaron  
  [MIT License](https://github.com/brandonaaron/jquery-mousewheel/blob/master/LICENSE.txt)

* **jQuery Outer HTML**
  by [Ca-Phun Ung](http://www.yelotofu.com/2008/08/jquery-outerhtml/)

* **[jquery-timing](https://github.com/creativecouple/jquery-timing)**  
  by [CreativeCouple](http://www.creativecouple.de/) - Peter Liske
  [MIT License](https://github.com/creativecouple/jquery-timing/blob/master/README.md)

* **[modernizr](www.modernizr.com)**  
  License: http://modernizr.com/license/

* **notextselect**  
  By [SLaks](http://stackoverflow.com/users/34397/slaks)

* **[-prefix-free](http://leaverou.github.com/prefixfree/)**  
  License: [MIT License](http://www.opensource.org/licenses/mit-license.php)
  Developed by [Lea Verou](http://lea.verou.me/)

* **resizeend**  
  By [zetce21](http://forum.jquery.com/user/zetce21)

* **[spin](http://fgnass.github.com/spin.js#v1.2.5)**  

* **[sprintf](http://www.diveintojavascript.com/projects/javascript-sprintf)**
  Copyright (c) Alexandru Marasteanu <alexaholic [at) gmail (dot] com>  
  All rights reserved.)  

* **[Tablesorter](http://tablesorter.com/docs/)**  
  License: http://opensource.org/licenses/mit-license.php
  Developed by [Christian Bach](https://twitter.com/lovepeacenukes)


Changelog
---------

### 3.3.0
* update js_init
* place generated less.css and coffee.js files inside the projects tmp-folder
* add tmpDir and tmpUrl to init args
* minor bugfixes

### pre 3.3
+ See changelog.md



Todo
----

+ Cleanup the js lib. (Bower?)
+ Convert ResponsiveImages and FileSelect into WP-Plugins.
+	build skeleton builder for standalones
+ Better documentation
+ Spell-check O_°

License
-------

Copyright (C) 2013 Hannes Diercks

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

changelog.md
THE MASTER CHANGELOG
====================

### 3.3.*
+   see [readme.md](https://github.com/Xiphe/THEMASTER/blob/master/readme.md)

### 3.2.7
* update spin.js
* add jquery.mousewheel.min.js
* add jquery.outer_html.min.js

### 3.2.6
* add validateCountFor, removeEmptyFrom, get_countErrorMessageFor & getValidCountStr methods to FileSelect.
* bugfix


### 3.2.5
* minor bugfixes

### 3.2.4
* *.inc.less files will be checked against updated files imported by the .inc.less file and re-parsed correctly
* minor bugfixes

### 3.2.3
* refined the js-var printing process.
* minor bugfixes.

### 3.2.2
* introducing xiphe_responsiveimages_availablehosts filter
* fileselect.js bugfixes

### 3.2.1
* new scrollend.js
* minified resizeend.js
* fixed bug where responsive images were not placed after generation.
* minor bugfixes

### 3.2
* exclude THEDEBUG, THETOOLS and THEWPTOOLS
* switch to composer libraries
* Autoloading of THEMASTER handled through composer
* Autoloading of other plugin and theme classes handled by THEAUTOLOADER
* include coffeescript parser

### 3.1.6
+ minor bugfixes
+ THEWPMASTER will register hooks now BEFORE less files are registered. To ensure error messages will be printed correctly
+ Shortend the merge of js globals

### 3.1.5
+ sanitizeValidation for fileSelect
+ bugfixes and restructuring of ResponsiveImages.

### 3.1.4
+ some File Select Bugs fixed
+ even better injection of less globals
+ URLNormalizer class for usage in THETOOLS::normalizeUrl() added 

### 3.1.3
+ Better injection of less variables
+ added THEWPMASTER::wrapJsVar()

### 3.1.2
+ updated jQuery UI themes
+ updated combobox.js
+ bugfixes for models and language in THEWPMASTER
+ added @masterRes and @baseUrl as less variables to every less file.

### 3.1.1
+ enhanced the content filter for ResponsiveImages.

### 3.1.0
+ new namespace autoloading
+ lots of small adjustments
+ update from 3.0.* needs many adjustments to depending themes.

### 3.0.12
+ Added option to disable CSSFix for a .less file. Just add "// NOFIX //" in the 5th or 6th line of your less file.
+ Internal enhancement of the THETOOLS::is_browser() method.
+ Added THETOOLS::get\_browser(), THETOOLS::get\_browserVersion() and THETOOLS::get\_layoutEngine()
+ Enhanced the slash methods in THETOOLS. They all now accept a boolean as second argument. When true the slashes in the passed string will be unified.
+ Deprecated THETOOLS::get\_cleanedPath() in favor of THETOOLS::unify\_slashes(). Same logic -> more semantic name.
+ Nicer ID for registered sources in THEWPMASTER::twpm_enqueue()
+ Added a > img { border: 0; } to base.less
+ Added and enhanced the gradient-mixins in elements.less

### 3.0.11
+ multiple methods can be bind to wordpress hook by using an array as value in the actions_ class variable.
+ Added .clear, hr.sep and hr.clear.sep as Buttons to TinyMCE

### 3.0.10
+ Introducing TM\WP() Method as shorthand for class_exists('\WP')
+ First standalone Bugfixes

### 3.0.9
+ Updated the PHPQuery helper.
+ Added settings_saved callback to THEWPSETTINGS.
+ bugfixes

### 3.0.8
+ FileSelect now tries to bind newly uploaded files to the parent post.
+ You can now change title and description of attachment files via the fileselect dialog.
+ Added THEWPTOOLS::get\_nav\_menu\_parent()
+ Bugfixes

### 3.0.7
+ FileSelect Class supports height and width for dimension validation.
+ Fixed bug with registered source destinations.
+ Added .nowrap class to base.less and tm-admin.less

### 3.0.6
+ Fixed some errors when project file is not found by THEWPBUILDER.
+ Added FileSelect::get_sizeStr() for easy comparing sizes of multiple files.
+ Fixed errors with deleted attachments in FileSelect.
+ Added ResponsiveImages::get_imagefile() for getting the full image path.
+ Added Percent values as possible width for ResponsiveImages
+ Added THEWPTOOLS::relUrl() for removing the wp-baseUrl from given url.
+ Fixed Bug in tm-fileselect.js where the selection of one field affected other siblings.

### 3.0.5
+ Added THETOOLS::shorten() for cutting a text after x letters 
+ Added THEWPTOOLS::get\_nav\_menu\_id()
+ Added THEWPTOOLS::get\_the\_date()
+ Bugfixes

### 3.0.4
+ Added DS, preDS, unDS and unPreDS to THETOOLS
+ Reactivated check\_folder\_structure_ in THEWPBUILDER used for gaining desired chmod on folders
  after project update.

### 3.0.3
+ Added FileSelect Class for manipulating the wordpress media upload dialog.
+ Added "attachment" as key for Setting type.
+ Added ResponsiveImage Class for preloading mini images and then get the right image files
  for the current screen-size.
+ Lots of bugfixes

### 3.0.2
+ Added THEWPTOOLS::posted_on() from TwentyEleven
+ LessPHP to 0.3.8 and fixed import of elements.less
+ fixed import of textdomain
+ Added edit\_theme\_options as capability for Theme Settings
+ Added 'textarea' and 'tiny_mce' as keys for Setting type
+ Improved tm-admin.js setting-save-logic to enable tinyMCE
+ Improved def Theme template.

### 3.0.1
+ Added THETOOLS::filter\_urlQuery() and THETOOLS::create\_sprite()
+ Fixed Names of Masterclass in THEBASE::extr() and THEBASE::get_view() methods.
+ Bigfixes.

### 3.0.0
+	Complete refactoring of the initiation process
+	Updates via own server now throu [WP Project Update API](https://github.com/Xiphe/WP-Project-Update-API)
+	added shorthand-options
+	merged !themasteroptions
+ deleted all deprecated methods.
+ THEDEBUG is now standalone.
+ Added THETOOLS for non-relating methods.
+ Added THEWPBUILDER as a skeleton builder for wp-plugins and wp-themes

### 2.1.1
+	added THEWPMASTER::twpm_wp_login() and his friends to enable &forceRederect=(true||'hash') in wordpress login url
+	added phpQuery Library 
	suport functions: THEMASTER::initPhpQuery(), THEMASTER::pq('<html...>') & THEMASTER::get_pqHtml()
+	added 'ipa' (iPad) as option for THEMASTER::is_browser()
+	added THESETTINGS::_set_setting($key, $value) for internal emergency setting changes
+	added auto-import of elements.less to each .less file

_themaster.php
<?php
/*
Plugin Name: THE MASTER
Plugin URI: https://github.com/Xiphe/THEMASTER
Namespace: Xiphe\THEMASTER
Description: The Master is here to enhance your Wordpress plug-ins and themes.
Version: 3.3.0
Date: 2013-05-08 10:31:39 +02.00
Author: Hannes Diercks aka Xiphe
Author URI: https://github.com/Xiphe
Update Server: http://wpupdates.xiphe.net/v1/
*/

/*
 Copyright (C) 2013 Hannes Diercks

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation; either version 2 of the License, or
 (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License along
 with this program; if not, write to the Free Software Foundation, Inc.,
 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
*/

namespace Xiphe\THEMASTER;

    /* -------------------- *
     *  DEVELOPMENMT STUFF  *
     * -------------------- */
    
    /*
     * In development i use one central symlinked version of this plugin.
     * Whenever something seems to be wrong i always uncomment this line
     * to check if i use the latest version or if the link was broken
     * through a sync process.
     */
    // die('THEMASTER SYMLINKED');

    /*
     * A quick breakpoint for xdebug.
     */ 
    // \xdebugBreak();

    require_once 'vendor/autoload.php';
    require_once 'vendor/xiphe/thedebug/globaldebug.php';

    /*
     * Some settings for development (if anything inside the master is broken).
     */
    // $tmTextID = basename(dirname(__FILE__)).'/'.basename(__FILE__);
    // $tmSettingsID = 'THEMASTER_' . strtoupper($tmTextID);
    // error_reporting(E_ALL);
    // ini_set('display_errors', 1);
    // define($tmSettingsID.'_ERRORREPORTING', true);
    // \Xiphe\THEDEBUG::$modus = 'FirePHP';
    // \Xiphe\THEDEBUG::enable();


/*
 * Define THEMASTER_HAS_WPRDPRESS for use in WP().
 */
if (class_exists('\WP')) {
    define('THEMASTER_HAS_WPRDPRESS', true);
} else {
    define('THEMASTER_HAS_WPRDPRESS', false);
}

/**
 * Shorthand funcion to check if Wordpress is available.
 *
 * @uses    const THEMASTER_HAS_WPRDPRESS
 * @return  bool  true if \WP Class exists - false if not.
 */
function WP()
{
    return THEMASTER_HAS_WPRDPRESS;
}

/*
 * I am using the DS constant as a shorthand for DIRECTORY_SEPARATOR.
 */
if (!defined('DS')) {
    define('DS', DIRECTORY_SEPARATOR);
} elseif (DS !== DIRECTORY_SEPARATOR) {
    $msg = __('The required constant "DS" is not available so !themaster will be deactivated.', 'themaster');
    if (WP()) {
        add_action(
            'admin_notices',
            function () use ($msg) {
                echo "<div class=\"error\"><p>$msg</p></div>";
            }
        );
        deactivate_plugins(__FILE__);
    } else {
        die($msg);
    }
}

/*
 * Register itself to be updatable.
 * This is the easyest way to enable the update logic of THEWPUPDATES
 * to a plugin that does not use !THE MASTER.
 */
$GLOBALS['Xiphe\THEMASTER\Updatable'][] = __FILE__;

/*
 * Save the path to this file.
 */
define('THEMASTER_PROJECTFILE', __FILE__);

/*
 * Save the path to this file.
 */
define('THEMASTER_PROJECTFOLDER', dirname(__FILE__).DS);

/*
 * Save the path to this file.
 */
define('THEMASTER_COREFOLDER', dirname(__FILE__).DS.'core'.DS);

/*
 * Register an autoloader for Themes and Plugins that do not have one.
 */
spl_autoload_register(array('Xiphe\THEMASTER\core\THEAUTOLOADER', 'autoload'));

if (WP()) {
    /**
     * Transforms the filepath as if it is a subdirectory of wp-content/themes or
     * wp-content/plugins.
     *
     * This is required for symliked projects because the activation_hooks of wordpress
     * do not work if the project is not located inside the wp-content folder.
     *
     * @param  string  $path     the realpath of the project
     * @param  boolean $isTheme  set true for themes
     * @param  boolean $hasNoDir set true if is one-file-plugin.
     * @return string            the symlink path.
     */
    function get_wpInstallPath($path, $isTheme = false, $hasNoDir = false)
    {
        $rel = !$hasNoDir ? basename(dirname($path)).DS : '';
        $rel .= basename($path);
        $rel = ABSPATH.'wp-content'.DS.(!$isTheme ? 'plugins' : 'themes').DS.$rel;
        return preg_replace('/[\\\|\/]/', DS, $rel);
    }

    /*
     * Register activation hook.
     */
    register_activation_hook(
        get_wpInstallPath(__FILE__),
        function () {
            core\THEWPMASTER::_masterActivate();
        }
    );

    /*
     * Register activation hook.
     */
    register_deactivation_hook(
        get_wpInstallPath(__FILE__),
        function () {
            core\THEWPMASTER::_masterDeactivate();
        }
    );
}


/*
 * Include core File - automaticaly includes required core files and instantiates a base instance.
 */
if (!defined('THEMINIWPMASTERAVAILABLE') && !defined('THEMINIMASTERAVAILABLE')) {
    try {
        if (!defined('THEMINIWPMASTERAVAILABLE') && WP()) {
            $GLOBALS['THEMINIWPMASTER'] = new core\THEWPMASTER('MINIMASTER');
            define('THEMINIWPMASTERAVAILABLE', true);
        } elseif (!defined('THEMINIMASTERAVAILABLE')) {
            $GLOBALS['THEMINIMASTER'] = new core\THEMASTER('MINIMASTER');
            define('THEMINIMASTERAVAILABLE', true);
        }
    } catch (\Exception $e) {
        /*
         * Errors Occured -> try to write an admin notice.
         */
        collect_tmInitErrors($e);
    }
}

/**
 * Basic Exception Handler.
 * 
 * @param  Exception $e Runtime Exception from MINIMASTER init.
 * @return void
 */
function collect_tmInitErrors($e) {
    if (!isset($GLOBALS['THEWPMASTERINITERRORS'])) {
        if (WP()) {
            $GLOBALS['THEWPMASTERINITERRORS'] = array();
            add_action('admin_notices', function() {
                foreach($GLOBALS['THEWPMASTERINITERRORS'] as $e) {
                    echo '<div class="error"><p>'.$e->getMessage().'<br />File:'.$e->getFile().' Line:'.$e->getLine().'</p></div>';
                }
            });
        } else {
            echo '<div class="error"><p>'.$e->getMessage().'<br />File:'.$e->getFile().' Line:'.$e->getLine().'</p></div>';
            return;
        }
    }
    $k = $e->getFile().$e->getLine();
    $GLOBALS['THEWPMASTERINITERRORS'][$k] = $e;
}

/**
 * Initiation for Plugins and Themes that want to use THE MASTER.
 * Automaticaly parses initiation args from main plugin file or 
 * theme style.
 *
 * @param   mixed  $initArgs optional additional initiation arguments. 
 *                           Keys are overwriting parsed init args.
 *                           Set Key "projectName" to prevent auto parsing.
 *                           This can also be a path to the projects info
 *                           file so the debug backtrace is not required
 *                           to get the called file.
 * @param   string $file     This can be a path to infofile if the first
 *                           param contains additional init args.    
 * @return  object           Instance of the Projects Master or false if error.
 */
function INIT($initArgs = null, $file = null)
{
    /*
     * If init args or key projectName is not set and given project-info-file is valid.
     */
    if ((!is_array($initArgs) || !isset($initArgs['projectName']))
     && (null === $initArgs
         || (is_string($initArgs) && file_exists($initArgs))
         || (is_string($file) && file_exists($file))
    )) {
        /*
         * Get initiation arguments from project file.
         */
        $filesInitArgs = core\THEWPBUILDER::get_initArgs($initArgs, $file);

        /*
         * Error occurred.
         */
        if (empty($filesInitArgs)) {
            return false;
        }

        /*
         * Check if additional args were given and merge them.
         */
        if (is_array($initArgs)) {
            $initArgs = array_merge($filesInitArgs, $initArgs);
        } else {
            $initArgs = $filesInitArgs;
        }
        
        /*
         * Add Masterflag.
         */
        $initArgs['isMaster'] = true;
    } 

    if (isset($initArgs['namespace']) && isset($initArgs['basePath'])) {
        core\THEAUTOLOADER::add($initArgs['namespace'], $initArgs['basePath']);
    }

    /*
     * Try to build a new Master with the initiation arguments.
     */
    try {
        $r = core\THEWPMASTER::sGet_instance('Master', $initArgs);
    } catch (\Exception $e) {
        /*
         * Errors Occured -> try to write an admin notice.
         */
        collect_tmInitErrors($e);

        /*
         * Do not return the object because it was not initated correctly.
         */
        $r = false;
    }
    return $r;
}

/**
 * This function can be used to automaticaly build skeletion projects.
 *
 * 1.  Check out `wp-content/plugins/_themaster/templates/new`.
 * 2.  Copy the theme or plugin folder into your wp-content/themes 
 *     or wp-content/plugins folder.
 * 3.  Rename the projects foldername
 * 3.b If you are generating a plugin also rename the plugin file.
 * 4.  Set the Project name in style.css or plugin file.
 * 5.  Activate the Project.
 * 6.  Revisit the style.css or project file and fill in additional informations.
 *
 * @param  bool   $extended set false if the project should just contain the absolute basic files and logic.
 * @param  string $template the name of the template that should be used. Default is "def" and it's the
 *                         only one that's delivered at the time of 3.0
 * @return void
 */
function BUILD($extended = true, $template = 'def')
{
    $args = core\THEWPBUILDER::get_initArgs();

    if (!isset($args['projectName'])) {
        throw new \Exception('BUILDTHEWPMASTERPLUGIN called in invalid file.', 1);
    } elseif (!isset($args['date'])) {
        core\THEWPBUILDER::sbuild('init', $args, $template, $extended);
    } elseif (!isset($args['version'])) {
        core\THEWPBUILDER::missing_initArgs($args);
    } else {
        core\THEWPBUILDER::sbuild('full', $args, $template, $extended);
    }
}
?>